hail/src/main/scala/is/hail/types/physical/stypes/concrete/SNDArraySlice.scala

```diff
@@ -0,0 +1,137 @@
+package is.hail.types.physical.stypes.concrete
+
+import is.hail.annotations.Region
+import is.hail.asm4s._
+import is.hail.expr.ir.{EmitCode, EmitCodeBuilder}
+import is.hail.types.physical.{PCanonicalNDArray, PNDArray, PNumeric, PPrimitive, PType}
+import is.hail.types.physical.stypes.{EmitType, SCode, SSettable, SType, SValue}
+import is.hail.types.physical.stypes.interfaces.{SBaseStructCode, SNDArray, SNDArrayCode, SNDArrayValue, primitive}
+import is.hail.types.physical.stypes.primitives.SInt64
+import is.hail.types.virtual.{TInt64, TNDArray, TTuple, Type}
+import is.hail.utils.{FastIndexedSeq, toRichIterable}
+
+case class SNDArraySlice(pType: PCanonicalNDArray) extends SNDArray {
+  def nDims: Int = pType.nDims
+
+  override def elementByteSize: Long = pType.elementType.byteSize
+
+  override def elementType: SType = pType.elementType.sType
+
+  override def elementPType: PType = pType.elementType
+
+  override lazy val virtualType: TNDArray = pType.virtualType
+
+  override def copiedType: SType = SNDArrayPointer(pType)
+
+  override def storageType(): PType = pType
+
+  override def castRename(t: Type): SType = SNDArrayPointer(pType.deepRename(t))
+
+  override def _coerceOrCopy(cb: EmitCodeBuilder, region: Value[Region], value: SCode, deepCopy: Boolean): SCode = ???
```

Comment:
implement this by matching on stype and returning self if the stype of `value` is equal to `this` and deepCopy is false, and throwing a match error otherwise.